### PR TITLE
Fixed entire api refactoring and started rework on rs api surface.

### PR DESCRIPTION
**Pull request by marchermans**
This should eventually fix the RS, work in progress.
Api reworked to make things unit testable....

### CODE REVIEW

File: src/api/java/com/minecolonies/coremod/entity/ai/util/StructureIterator.java [hunk 1]
```diff
@@ -158,7 +157,7 @@ public boolean doesStructureBlockEqualWorldBlock()
             }
             else if ((structureBlock instanceof BlockStairs && structureBlockState.equals(worldBlockState))
                        || BlockUtils.isGrassOrDirt(structureBlock, worldBlock, structureBlockState, worldBlockState)
-                       || (worldBlock == ModBlocks.blockRack && BlockMinecoloniesRack.shouldBlockBeReplacedWithRack(structureBlock)))
+                       || (worldBlock instanceof IBlockMinecoloniesRack && IBlockMinecoloniesRack.shouldBlockBeReplacedWithRack(structureBlock)))
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/coremod/entity/ai/util/StructureIterator.java]: Isn't that less efficient?

Comment (marchermans) [reply src/api/java/com/minecolonies/coremod/entity/ai/util/StructureIterator.java]: Not really.
The difference is not notable.

It allows for easier extension in the future as well.

Comment (someaddons) [reply src/api/java/com/minecolonies/coremod/entity/ai/util/StructureIterator.java]: afaik instanceof is pretty fast these days

**Comment by Raycoms** [comment]
kk

Am Fr., 26. Juli 2019 um 13:46 Uhr schrieb someaddons <
notifications@github.com>:

**Pull request closed without merging**